prov/gni/src/gnix_buddy_allocator.c

```diff
@@ -0,0 +1,353 @@
+
+		*block_size *= 2;
+
+		_gnix_clear_bit(&alloc_handle->bitmap,
+				BITMAP_INDEX(*ptr, alloc_handle->base,
+					     MIN_BLOCK_SIZE, *block_size));
+	}
+
+}
+
+int _gnix_buddy_allocator_create(void *base, size_t len, size_t max,
+				 handle_t **alloc_handle)
+{
+	char err_buf[256] = {0}, *error = NULL;
+	int fi_errno;
+
+	GNIX_TRACE(FI_LOG_EP_CTRL, "\n");
+
+	/* Ensure parameters are valid */
+	if (!base || !len || !max || max > len || !alloc_handle ||
+	    IS_NOT_POW_TWO(max) || (len % max) || !(len / MIN_BLOCK_SIZE * 2)) {
+
+		GNIX_WARN(FI_LOG_EP_CTRL,
+			  "Invalid parameter to _gnix_buddy_allocator_create."
+			  "\n");
+		return -FI_EINVAL;
+	}
+
+	*alloc_handle = calloc(1, sizeof(handle_t));
+
+	if (!alloc_handle) {
+		error = strerror_r(errno, err_buf, sizeof(err_buf));
+		GNIX_WARN(FI_LOG_EP_CTRL,
+			  "Could not create buddy allocator handle.",
+			  error);
+		return -FI_ENOMEM;
+	}
+
+	fastlock_init(&alloc_handle[0]->lock);
+	fastlock_acquire(&alloc_handle[0]->lock);
+
+	alloc_handle[0]->base = base;
+	alloc_handle[0]->len = len;
+	alloc_handle[0]->max = max;
+
+	if (__gnix_buddy_create_lists(alloc_handle[0])) {
+		fastlock_release(&alloc_handle[0]->lock);
+		free(*alloc_handle);
+		return -FI_ENOMEM;
+	}
+
+	if ((fi_errno = _gnix_alloc_bitmap(&alloc_handle[0]->bitmap,
+					   len / MIN_BLOCK_SIZE * 2))) {
+
+		free(&alloc_handle[0]->lists);
+		free(*alloc_handle);
+	}
+
+	fastlock_release(&alloc_handle[0]->lock);
+
+	return fi_errno;
+}
+
+int _gnix_buddy_allocator_destroy(handle_t *alloc_handle)
+{
+	GNIX_TRACE(FI_LOG_EP_CTRL, "\n");
+
+	if (!alloc_handle) {
+		GNIX_WARN(FI_LOG_EP_CTRL,
+			  "Invalid parameter to _gnix_buddy_allocator_destroy."
+			  "\n");
+		return -FI_EINVAL;
+	}
+
+	fastlock_acquire(&alloc_handle->lock);
+
+	free(alloc_handle->lists);
+
+	while (_gnix_free_bitmap(&alloc_handle->bitmap)) {
+		GNIX_WARN(FI_LOG_EP_CTRL,
+			  "Trying to free buddy allocator handle bitmap.");
+		sleep(1);
+	}
+
+	fastlock_release(&alloc_handle->lock);
+	fastlock_destroy(&alloc_handle->lock);
+
+	free(alloc_handle);
+
+	return FI_SUCCESS;
+}
+
+int _gnix_buddy_alloc(handle_t *alloc_handle, void **ptr, size_t len)
+{
+	size_t block_size, i = 0;
+
+	GNIX_TRACE(FI_LOG_EP_CTRL, "\n");
+
+	if (!alloc_handle || !ptr || !len || len > alloc_handle->max) {
+		GNIX_WARN(FI_LOG_EP_CTRL,
+			  "Invalid parameter to _gnix_buddy_alloc.\n");
+		return -FI_EINVAL;
+	}
+
+	block_size = BLOCK_SIZE(len, MIN_BLOCK_SIZE);
+	i = (size_t) LIST_INDEX(block_size, MIN_BLOCK_SIZE);
+
+	fastlock_acquire(&alloc_handle->lock);
+
+	if (dlist_empty(alloc_handle->lists + i) &&
+	    __gnix_buddy_find_block(alloc_handle, i)) {
+
+		fastlock_release(&alloc_handle->lock);
+		GNIX_WARN(FI_LOG_EP_CTRL,
+			  "Could not allocate buddy block.\n");
+		return -FI_ENOMEM;
+	}
+
+	/*
+	 * Remove the block from opposite sides of adjacent lists to reduce
+	 * fragmentation and improve coalescing
+	 */
+	if (i % 2) {
+		dlist_remove(*ptr = alloc_handle->lists[i].prev);
```

Comment:
That works.  I agree.

> On Jan 7, 2016, at 6:13 PM, James Swaro <notifications@github.com> wrote:
> 
> In prov/gni/src/gnix_buddy_allocator.c:
> 
> > +
> > +	if (dlist_empty(alloc_handle->lists + i) &&
> > +	    __gnix_buddy_find_block(alloc_handle, i)) {
> > +
> > +		fastlock_release(&alloc_handle->lock);
> > +		GNIX_WARN(FI_LOG_EP_CTRL,
> > +			  "Could not allocate buddy block.\n");
> > +		return -FI_ENOMEM;
> > +	}
> > +
> > +	/*
> > +	 * Remove the block from opposite sides of adjacent lists to reduce
> > +	 * fragmentation and improve coalescing
> > +	 */
> > +	if (i % 2) {
> > +		dlist_remove(*ptr = alloc_handle->lists[i].prev);
> While we are on the topic of funky looking things, you could do this to avoid the conditional ... but don't
> 
> ptr = (struct dlist_entry *) ((struct dlist_entry *) &alloc_handle->lists[i]) + (i & 1));
> 
> readability is far more important than that.
> 
> —
> Reply to this email directly or view it on GitHub.
>